lib/Dialect/FIRRTL/Transforms/ModuleSummary.cpp

```diff
@@ -0,0 +1,99 @@
+//===- ModuleSummaryt.cpp ---------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "mlir/IR/Threading.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+struct ModuleSummaryPass : public ModuleSummaryBase<ModuleSummaryPass> {
+  struct KeyTy {
+    SmallVector<size_t> portSizes;
+    size_t opcount;
+    bool operator==(const KeyTy &rhs) const {
+      return portSizes == rhs.portSizes && opcount == rhs.opcount;
+    }
+  };
+
+  size_t countOps(FModuleOp mod) {
+    size_t retval = 0;
+    mod.walk([&](Operation *op) { retval += 1; });
+    return retval;
+  }
+
+  SmallVector<size_t> protSig(FModuleOp mod) {
+    SmallVector<size_t> ports;
+    for (auto p : mod.getPortTypes())
+      ports.push_back(cast<FIRRTLBaseType>(cast<TypeAttr>(p).getValue())
+                          .getBitWidthOrSentinel());
+    return ports;
+  }
+
+  void runOnOperation() override;
+};
+} // namespace
+
+namespace mlir {
+// NOLINTNEXTLINE(readability-identifier-naming)
+inline llvm::hash_code hash_value(const ModuleSummaryPass::KeyTy &element) {
+  return llvm::hash_combine(element.portSizes.size(),
+                            llvm::hash_combine_range(element.portSizes.begin(),
+                                                     element.portSizes.end()),
+                            element.opcount);
+}
+} // namespace mlir
+
+namespace llvm {
+// Type hash just like pointers.
+template <>
+struct DenseMapInfo<ModuleSummaryPass::KeyTy> {
+  using KeyTy = ModuleSummaryPass::KeyTy;
+  static KeyTy getEmptyKey() { return {{}, ~0ULL}; }
+  static KeyTy getTombstoneKey() { return {{}, ~0ULL - 1}; }
+  static unsigned getHashValue(KeyTy val) { return mlir::hash_value(val); }
+  static bool isEqual(KeyTy LHS, KeyTy RHS) { return LHS == RHS; }
+};
+} // namespace llvm
+
+void ModuleSummaryPass::runOnOperation() {
+  auto circuit = getOperation();
+
+  using MapTy = DenseMap<KeyTy, SmallVector<FModuleOp>>;
+  MapTy data;
+
+  for (auto mod : circuit.getOps<FModuleOp>()) {
+    auto p = protSig(mod);
+    auto n = countOps(mod);
+    data[{p, n}].push_back(mod);
```

Comment:
Nit:
Seems a bit heavy to have SmallVector's as keys in a DenseMap, especially since it's only used to group by hash not actually important to have the vector data itself.  Could have keys be port-size-hash plus op count or something.

Since a debugging pass maybe not worth worrying about but just a thought..

Reply:
The key is the vector of the bitwidth of each port.  so differences in port widths or differences in number of ports should both cause the hash to be different.

Reply:
Right! That's why it'd be port-size-hash and not just port-size as part of the key.  We only need the actual vector to get enough to know which summary entry to bin this with, don't really need the vector itself if that makes sense.

However having it in the key helps with equality-checking beyond hash-equivalence and welp anyway not sure how to slot this in neatly and if it matters we'll know and probably other things will fall over first anyway.

Marking resolved, thanks for the response.